File: tests/DesktopGrass.Native.Tests/src/pine_tests.cpp
// pine_tests.cpp - §15.1 Winter pine trees (slot-bound, mirrors §14 cacti).

#include "TestHelpers.h"
#include "Sim.h"
#include "snapshot_data.h"

#include <cmath>
#include <cstddef>
#include <vector>

using namespace desktopgrass;

namespace {

constexpr double kMonitor1920 = 1920.0;

struct ExpectedTree {
    std::size_t slotIndex = 0;
    uint8_t variant = 0;
    double height = 0.0;
    double width = 0.0;
    int tierCount = 0;
};

ExpectedTree first_expected_tree(std::size_t bladeCount) {
    Prng p;
    prng_init(p, CANONICAL_TEST_SEED ^ PINE_PRNG_SALT);

    for (std::size_t i = 0; i < bladeCount; ++i) {
        const double r = prng_uniform(p, 0.0, 1.0);
        if (r >= PINE_PROBABILITY) continue;

        ExpectedTree expected{};
        expected.slotIndex = i;
        const double variantDraw = prng_uniform(p, 0.0, 1.0);
        expected.variant = variantDraw < BIRCH_VARIANT_PROBABILITY ? 1 : 0;
        expected.height = prng_uniform(p, PINE_HEIGHT_MIN, PINE_HEIGHT_MAX);
        if (expected.variant == 1) {
            expected.width = prng_uniform(p, BIRCH_TRUNK_WIDTH_MIN, BIRCH_TRUNK_WIDTH_MAX);
        } else {
            expected.width = prng_uniform(p, PINE_WIDTH_MIN, PINE_WIDTH_MAX);
        }
        const double tierDraw = prng_uniform(p,
            static_cast<double>(PINE_TIER_COUNT_MIN),
            static_cast<double>(PINE_TIER_COUNT_MAX + 1));
        int tiers = static_cast<int>(std::floor(tierDraw));
        if (tiers < PINE_TIER_COUNT_MIN) tiers = PINE_TIER_COUNT_MIN;
        if (tiers > PINE_TIER_COUNT_MAX) tiers = PINE_TIER_COUNT_MAX;
        expected.tierCount = tiers;
        return expected;
    }

    Assert::Fail(L"Canonical seed produced no tree slot", LINE_INFO());
    return {};
}

} // anonymous

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(PineTests)
{
public:
TEST_METHOD(PineConstantsArePinned) {
    Assert::IsTrue(PINE_PROBABILITY == Near(0.0075));
    Assert::IsTrue(PINE_HEIGHT_MIN == Near(45.0));
    Assert::IsTrue(PINE_HEIGHT_MAX == Near(90.0));
    Assert::IsTrue(PINE_WIDTH_MIN  == Near(28.0));
    Assert::IsTrue(PINE_WIDTH_MAX  == Near(48.0));
    Assert::IsTrue(PINE_TIER_COUNT_MIN == 2);
    Assert::IsTrue(PINE_TIER_COUNT_MAX == 4);
    Assert::IsTrue(PINE_TIP_TAPER == Near(0.25));
    Assert::IsTrue(PINE_TIER_OVERLAP == Near(0.15));
    Assert::IsTrue(PINE_SNOW_CAP_FRACTION == Near(0.30));
    Assert::IsTrue(PINE_COLOR == 0xFF1B5E20u);
    Assert::IsTrue(PINE_PRNG_SALT == 0x50494E4550494E45ull);
}

TEST_METHOD(BirchConstantsArePinned) {
    Assert::IsTrue(BIRCH_VARIANT_PROBABILITY == Near(0.30));
    Assert::IsTrue(BIRCH_TRUNK_WIDTH_MIN == Near(4.0));
    Assert::IsTrue(BIRCH_TRUNK_WIDTH_MAX == Near(7.0));
    Assert::IsTrue(BIRCH_BARK_MARK_COUNT == 5);
    Assert::IsTrue(BIRCH_BARK_MARK_LENGTH_FRAC == Near(0.50));
    Assert::IsTrue(BIRCH_BRANCH_COUNT == 6);
    Assert::IsTrue(BIRCH_SNOW_CAP_FRACTION == Near(0.18));
    Assert::IsTrue(BIRCH_BARK_COLOR == 0xFFEFEFE6u);
    Assert::IsTrue(BIRCH_MARK_COLOR == 0xFF2A2A28u);
}

TEST_METHOD(SimSetSceneWinterPromotesSomeSlotsToTrees) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Winter);

    Assert::IsTrue(sim.currentScene == Scene::Winter);
    std::size_t treeCount = 0;
    for (const Blade& b : sim.blades) {
        if (b.isPine) {
            ++treeCount;
            Assert::IsTrue(b.pineTierCount >= PINE_TIER_COUNT_MIN);
            Assert::IsTrue(b.pineTierCount <= PINE_TIER_COUNT_MAX);
            Assert::IsTrue(b.pineHeight >= PINE_HEIGHT_MIN);
            Assert::IsTrue(b.pineHeight <= PINE_HEIGHT_MAX);
            const double widthMin = (b.treeVariant == 1) ? BIRCH_TRUNK_WIDTH_MIN : PINE_WIDTH_MIN;
            const double widthMax = (b.treeVariant == 1) ? BIRCH_TRUNK_WIDTH_MAX : PINE_WIDTH_MAX;
            Assert::IsTrue(b.pineWidth >= widthMin);
            Assert::IsTrue(b.pineWidth <= widthMax);
        }
    }
    Assert::IsTrue(treeCount >= 1);
    Assert::IsTrue(treeCount <= 25);
}

TEST_METHOD(FirstTreeMatchesTheSpecDerivedPRNGSnapshot) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    const ExpectedTree expected = first_expected_tree(sim.blades.size());

    sim_set_scene(sim, Scene::Winter);

    Assert::IsTrue(expected.slotIndex < sim.blades.size());
    const Blade& b = sim.blades[expected.slotIndex];
    Assert::IsTrue(b.isPine);
    Assert::IsTrue(b.treeVariant == expected.variant);
    Assert::IsTrue(b.pineHeight == Near(expected.height).margin(1e-12));
    Assert::IsTrue(b.pineWidth  == Near(expected.width).margin(1e-12));
    Assert::IsTrue(b.pineTierCount == expected.tierCount);
}

TEST_METHOD(GrassSceneRestoresTreeSlotsToVanillaVariants) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    const ExpectedTree expected = first_expected_tree(sim.blades.size());
    Assert::IsTrue(expected.slotIndex < sim.blades.size());

    Blade& target = sim.blades[expected.slotIndex];
    target.isFlower = true;
    target.isMushroom = true;
    target.originalIsFlower = true;
    target.originalIsMushroom = true;

    sim_set_scene(sim, Scene::Winter);
    Assert::IsTrue(sim.blades[expected.slotIndex].isPine);
    Assert::IsFalse(sim.blades[expected.slotIndex].isFlower);
    Assert::IsFalse(sim.blades[expected.slotIndex].isMushroom);

    sim_set_scene(sim, Scene::Grass);
    Assert::IsFalse(sim.blades[expected.slotIndex].isPine);
    Assert::IsTrue(sim.blades[expected.slotIndex].treeVariant == 0);
    Assert::IsTrue(sim.blades[expected.slotIndex].isFlower);
    Assert::IsTrue(sim.blades[expected.slotIndex].isMushroom);
}

TEST_METHOD(WinterProducesBothPineAndBirchVariantsOverCanonicalSeed) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Winter);

    std::size_t pineCount = 0;
    std::size_t birchCount = 0;
    for (const Blade& b : sim.blades) {
        if (!b.isPine) continue;
        if (b.treeVariant == 0) {
            ++pineCount;
            Assert::IsTrue(b.pineWidth >= PINE_WIDTH_MIN);
            Assert::IsTrue(b.pineWidth <= PINE_WIDTH_MAX);
        } else {
            Assert::IsTrue(b.treeVariant == 1);
            ++birchCount;
            Assert::IsTrue(b.pineWidth >= BIRCH_TRUNK_WIDTH_MIN);
            Assert::IsTrue(b.pineWidth <= BIRCH_TRUNK_WIDTH_MAX);
        }
    }
    Assert::IsTrue(pineCount >= 1);
    Assert::IsTrue(birchCount >= 1);
}

TEST_METHOD(WinterSceneSuppressesMushroomsOnEverySlot) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    // Pre-mark a handful of slots as mushrooms; Winter must clear them all.
    for (std::size_t i = 0; i < sim.blades.size(); i += 17) {
        sim.blades[i].isMushroom = true;
        sim.blades[i].originalIsMushroom = true;
    }

    sim_set_scene(sim, Scene::Winter);

    for (const Blade& b : sim.blades) Assert::IsFalse(b.isMushroom);

    // Switching back to Grass must restore the original mushroom flags.
    sim_set_scene(sim, Scene::Grass);
    Assert::IsTrue(sim.blades[0].isMushroom == sim.blades[0].originalIsMushroom);
}

TEST_METHOD(WinterGrassHeightScaleIsPinned) {
    Assert::IsTrue(WINTER_GRASS_HEIGHT_SCALE == Near(0.5));
}

TEST_METHOD(TreeDepthConstantsArePinned) {
    Assert::IsTrue(TREE_BACKGROUND_PROBABILITY == Near(0.45));
    Assert::IsTrue(TREE_BG_SCALE == Near(0.62));
    Assert::IsTrue(TREE_BG_OPACITY == Near(0.78f));
}

TEST_METHOD(WinterMixesForegroundAndBackgroundTrees) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Winter);

    std::size_t fg = 0;
    std::size_t bg = 0;
    for (const Blade& b : sim.blades) {
        if (!b.isPine) continue;
        if (b.treeBackground) ++bg; else ++fg;
    }
    Assert::IsTrue(fg >= 1);
    Assert::IsTrue(bg >= 1);
}

TEST_METHOD(TreeDepthAssignmentIsDeterministicAcrossReEntry) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Winter);

    std::vector<bool> firstPass;
    for (const Blade& b : sim.blades) {
        if (b.isPine) firstPass.push_back(b.treeBackground);
    }

    // Leaving and re-entering Winter must reproduce the same depth layout.
    sim_set_scene(sim, Scene::Grass);
    sim_set_scene(sim, Scene::Winter);

    std::size_t idx = 0;
    for (const Blade& b : sim.blades) {
        if (!b.isPine) continue;
        Assert::IsTrue(idx < firstPass.size());
        Assert::IsTrue(b.treeBackground == firstPass[idx]);
        ++idx;
    }
    Assert::IsTrue(idx == firstPass.size());
}

TEST_METHOD(NonWinterScenesClearTheTreeBackgroundFlag) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Winter);
    sim_set_scene(sim, Scene::Grass);
    for (const Blade& b : sim.blades) Assert::IsFalse(b.treeBackground);
}

TEST_METHOD(WinterSceneLeavesTheCanonicalFirstBladeGeometryBitIdentical) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, kMonitor1920, 1.0);
    Assert::IsTrue(sim.blades.size() == desktopgrass::test::CANONICAL_BLADE_COUNT);

    sim_set_scene(sim, Scene::Winter);

    const Blade& first = sim.blades[0];
    const auto& expected = desktopgrass::test::CANONICAL_FIRST_10[0];
    Assert::IsTrue(first.baseX == Near(expected.baseX).margin(1e-12));
    Assert::IsTrue(first.height == Near(expected.height).margin(1e-12));
    Assert::IsTrue(first.thickness == Near(expected.thickness).margin(1e-12));
    Assert::IsTrue(first.hue == expected.hue);
    Assert::IsTrue(first.swayPhaseOffset == Near(expected.sway).margin(1e-12));
    Assert::IsTrue(first.stiffness == Near(expected.stiffness).margin(1e-12));
}
};
}

File: tests/DesktopGrass.Win2D.Tests/SimTests/PineTests.cs
// PineTests.cs - §15.1 Winter pine trees (slot-bound, mirrors §14 cacti).

using System;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class PineTests
{
    private const double Monitor1920 = 1920.0;

    private readonly struct ExpectedTree
    {
        public readonly int SlotIndex;
        public readonly byte Variant;
        public readonly double Height;
        public readonly double Width;
        public readonly int TierCount;

        public ExpectedTree(int slotIndex, byte variant, double height, double width, int tierCount)
        {
            SlotIndex = slotIndex;
            Variant = variant;
            Height = height;
            Width = width;
            TierCount = tierCount;
        }
    }

    private static Sim BuildSim(double density = Constants.DEFAULT_DENSITY)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, Monitor1920, density),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(Constants.CANONICAL_TEST_SEED, Monitor1920);
        sim.ResetEntities(Constants.CANONICAL_TEST_SEED);
        return sim;
    }

    private static ExpectedTree FirstExpectedTree(int bladeCount)
    {
        var prng = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.PINE_PRNG_SALT);
        for (int i = 0; i < bladeCount; i++)
        {
            double r = prng.Uniform(0.0, 1.0);
            if (r >= Constants.PINE_PROBABILITY) continue;

            double variantDraw = prng.Uniform(0.0, 1.0);
            byte variant = (variantDraw < Constants.BIRCH_VARIANT_PROBABILITY) ? (byte)1 : (byte)0;
            double height = prng.Uniform(Constants.PINE_HEIGHT_MIN, Constants.PINE_HEIGHT_MAX);
            double width = (variant == 1)
                ? prng.Uniform(Constants.BIRCH_TRUNK_WIDTH_MIN, Constants.BIRCH_TRUNK_WIDTH_MAX)
                : prng.Uniform(Constants.PINE_WIDTH_MIN, Constants.PINE_WIDTH_MAX);
            double tierDraw = prng.Uniform(Constants.PINE_TIER_COUNT_MIN, Constants.PINE_TIER_COUNT_MAX + 1);
            int tiers = (int)Math.Floor(tierDraw);
            if (tiers < Constants.PINE_TIER_COUNT_MIN) tiers = Constants.PINE_TIER_COUNT_MIN;
            if (tiers > Constants.PINE_TIER_COUNT_MAX) tiers = Constants.PINE_TIER_COUNT_MAX;
            return new ExpectedTree(i, variant, height, width, tiers);
        }
        throw new InvalidOperationException("canonical seed produced no tree slot");
    }

    [Fact]
    public void PineConstantsArePinned()
    {
        Assert.Equal(0.0075, Constants.PINE_PROBABILITY);
        Assert.Equal(45.0, Constants.PINE_HEIGHT_MIN);
        Assert.Equal(90.0, Constants.PINE_HEIGHT_MAX);
        Assert.Equal(16.0, Constants.PINE_WIDTH_MIN);
        Assert.Equal(28.0, Constants.PINE_WIDTH_MAX);
        Assert.Equal(2, Constants.PINE_TIER_COUNT_MIN);
        Assert.Equal(4, Constants.PINE_TIER_COUNT_MAX);
        Assert.Equal(0.25, Constants.PINE_TIP_TAPER);
        Assert.Equal(0.15, Constants.PINE_TIER_OVERLAP);
        Assert.Equal(0.30, Constants.PINE_SNOW_CAP_FRACTION);
        Assert.Equal(0xFF1B5E20u, Constants.PINE_COLOR);
        Assert.Equal(0x50494E4550494E45ul, Constants.PINE_PRNG_SALT);
    }

    [Fact]
    public void BirchConstantsArePinned()
    {
        Assert.Equal(0.30, Constants.BIRCH_VARIANT_PROBABILITY);
        Assert.Equal(4.0, Constants.BIRCH_TRUNK_WIDTH_MIN);
        Assert.Equal(7.0, Constants.BIRCH_TRUNK_WIDTH_MAX);
        Assert.Equal(5, Constants.BIRCH_BARK_MARK_COUNT);
        Assert.Equal(0.50, Constants.BIRCH_BARK_MARK_LENGTH_FRAC);
        Assert.Equal(6, Constants.BIRCH_BRANCH_COUNT);
        Assert.Equal(0.18, Constants.BIRCH_SNOW_CAP_FRACTION);
        Assert.Equal(0xFFEFEFE6u, Constants.BIRCH_BARK_COLOR);
        Assert.Equal(0xFF2A2A28u, Constants.BIRCH_MARK_COLOR);
    }

    [Fact]
    public void SetSceneWinterPromotesSomeSlotsToTrees()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Winter);

        int treeCount = 0;
        foreach (var b in sim.Blades)
        {
            if (b.IsPine)
            {
                treeCount++;
                Assert.InRange(b.PineTierCount, Constants.PINE_TIER_COUNT_MIN, Constants.PINE_TIER_COUNT_MAX);
                Assert.InRange(b.PineHeight, Constants.PINE_HEIGHT_MIN, Constants.PINE_HEIGHT_MAX);
                double widthMin = (b.TreeVariant == 1) ? Constants.BIRCH_TRUNK_WIDTH_MIN : Constants.PINE_WIDTH_MIN;
                double widthMax = (b.TreeVariant == 1) ? Constants.BIRCH_TRUNK_WIDTH_MAX : Constants.PINE_WIDTH_MAX;
                Assert.InRange(b.PineWidth, widthMin, widthMax);
            }
        }
        Assert.InRange(treeCount, 1, 25);
    }

    [Fact]
    public void FirstTreeMatchesSpecDerivedPrngSnapshot()
    {
        var sim = BuildSim();
        var expected = FirstExpectedTree(sim.Blades.Length);

        sim.SetScene(Scene.Winter);

        Assert.InRange(expected.SlotIndex, 0, sim.Blades.Length - 1);
        var b = sim.Blades[expected.SlotIndex];
        Assert.True(b.IsPine);
        Assert.Equal(expected.Variant, b.TreeVariant);
        Assert.Equal(expected.Height, b.PineHeight, 12);
        Assert.Equal(expected.Width, b.PineWidth, 12);
        Assert.Equal(expected.TierCount, b.PineTierCount);
    }

    [Fact]
    public void GrassSceneRestoresTreeSlotsToVanillaVariants()
    {
        var sim = BuildSim();
        var expected = FirstExpectedTree(sim.Blades.Length);
        Assert.InRange(expected.SlotIndex, 0, sim.Blades.Length - 1);

        sim.Blades[expected.SlotIndex].IsFlower = true;
        sim.Blades[expected.SlotIndex].IsMushroom = true;
        sim.Blades[expected.SlotIndex].OriginalIsFlower = true;
        sim.Blades[expected.SlotIndex].OriginalIsMushroom = true;

        sim.SetScene(Scene.Winter);
        Assert.True(sim.Blades[expected.SlotIndex].IsPine);
        Assert.False(sim.Blades[expected.SlotIndex].IsFlower);
        Assert.False(sim.Blades[expected.SlotIndex].IsMushroom);

        sim.SetScene(Scene.Grass);
        Assert.False(sim.Blades[expected.SlotIndex].IsPine);
        Assert.Equal((byte)0, sim.Blades[expected.SlotIndex].TreeVariant);
        Assert.True(sim.Blades[expected.SlotIndex].IsFlower);
        Assert.True(sim.Blades[expected.SlotIndex].IsMushroom);
    }

    [Fact]
    public void WinterProducesBothPineAndBirchVariants()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Winter);

        int pineCount = 0;
        int birchCount = 0;
        foreach (var b in sim.Blades)
        {
            if (!b.IsPine) continue;
            if (b.TreeVariant == 0) pineCount++;
            else { Assert.Equal((byte)1, b.TreeVariant); birchCount++; }
        }
        Assert.True(pineCount >= 1, $"expected >=1 pine, got {pineCount}");
        Assert.True(birchCount >= 1, $"expected >=1 birch, got {birchCount}");
    }

    [Fact]
    public void WinterSceneSuppressesMushroomsOnEverySlot()
    {
        var sim = BuildSim();
        for (int i = 0; i < sim.Blades.Length; i += 17)
        {
            sim.Blades[i].IsMushroom = true;
            sim.Blades[i].OriginalIsMushroom = true;
        }

        sim.SetScene(Scene.Winter);

        foreach (var b in sim.Blades) Assert.False(b.IsMushroom);

        sim.SetScene(Scene.Grass);
        Assert.Equal(sim.Blades[0].OriginalIsMushroom, sim.Blades[0].IsMushroom);
    }

    [Fact]
    public void WinterGrassHeightScaleIsPinned()
    {
        Assert.Equal(0.5, Constants.WINTER_GRASS_HEIGHT_SCALE);
    }

    [Fact]
    public void WinterSceneLeavesCanonicalFirstBladeBitIdentical()
    {
        var sim = BuildSim(density: 1.0);
        var before = sim.Blades[0];

        sim.SetScene(Scene.Winter);

        var after = sim.Blades[0];
        Assert.Equal(before.BaseX, after.BaseX, 12);
        Assert.Equal(before.Height, after.Height, 12);
        Assert.Equal(before.Thickness, after.Thickness, 12);
        Assert.Equal(before.Hue, after.Hue);
        Assert.Equal(before.SwayPhaseOffset, after.SwayPhaseOffset, 12);
        Assert.Equal(before.Stiffness, after.Stiffness, 12);
    }
}
